stackify, style changes, backlog

fredefox opened the pull request:
I've added a backlog with comments from our meeting today.

I've also added *all of GF* as a git submodule dependency so that new devs can build this project out of the box.

I've added qualified imports in relevant places in `muste-lib` so we can see what is needed in the other components. Specifically I had a look at yanking out labeled trees (`LTree`s). Please see the relevant comment in the Backlog.

heatherleaf commented on BACKLOG.md:
```diff
@@ -0,0 +1,38 @@
+-----------------
+
+I consider removing the haste front end and `Main.hs` in `muste-lib`. I
+don't know if this is needed for some special purpose.
```
I agree: but after the refactorisation, perhaps consider adding offline tests

heatherleaf replied:
There is a Test directoy, but it's quite old and obsolete. Perhaps it's easier to scrap it and start over with testing when the refactoring is mostly finished.

heatherleaf commented on BACKLOG.md:
```diff
@@ -0,0 +1,38 @@
+
+General refactoring
+-------------------
+There are 3 representation of a tree.
```
5 actually (at least): TTree, LTree, MetaTTree, CostTree, GFAbsTree

GFAbsTree is used by PGF, but should only be used in an PGF interface module (e.g., Tree/Internal.hs). Currently it's also used in Main.hs

heatherleaf commented on muste-lib/Main.hs:
```diff
@@ -1,15 +1,32 @@
+-- | I think this is some kind of command line interface for the
```
I think this file can be scrapped (and replaced by tests in the future)

heatherleaf commented on BACKLOG.md:
```diff
@@ -0,0 +1,38 @@
+Remove `LTree`. `LTree` is currently not exported from `muste-lib`, it
+is however used internally. E.g. in `linearizeTree` which is used in
+the ajax-backend.
+
```
There is also LinToken and Linearization (in Muste.hs) - should be merged into one.

heatherleaf replied:
Which functions fit best where? 

- linearization functions can be moved to a new module (e.g. Muste.Linearize): LinToken, Linearization, linearizeTree (are defined in Muste.hs)

- the Ajax backend should only refer to functions exported by Muste.hs: now Ajax.hs imports Muste.Tree, Main-CGI.hs imports Muste.Grammar, and Database.hs and Protocol.hs imports both Muste.Tree, Muste.Grammar and PGF

- perhaps Database.hs should be moved to the muste-backend?? Also, Database.hs contains explicit example trees

- is there a need for Muste.Grammar.Internal? and Muste.Tree.Internal? 

- only Muste.Grammar (or Muste.Grammar.Internal) should use the PGF module -- everything that the other modules need should be exported by Muste.Grammar

- Muste.Feat currently defines the types Grammar, Rule, FunType -- this should be done by Muste.Grammar (and the functions getAllRules, getRuleType)

- Muste.Feat currently defines TTree -- this should be in Muste.Tree instead

- functions in Muste.Prune that should be in other files:
   - getFunctions, sameRoot, getMetas, treeDiff, treeIsRecursive, ... (move to Muste.Tree)
   - editDistance, isSubList, areDisjoint, noDuplicates, ... (move to Muste.Common)

fredefox commented:
Status update
=============

I have moved things around quite a lot.  There are now two packages
`muste` (previously `muste-lib`) and `muste-ajax` placed in
directories named accordingly.

Both directories follow the same structure.  There is a sub directory
called `src/` housing the (internal and external) modules for that
package.  There is also a `test/` directory (I haven't touched that so
much yet).  The package `muste-ajax` define two executables named
`muste-ajax` and `muste-cgi`.  These can be built independently (for
build instructions please refer to `README.md`).

Most modules in `muste` continue to have an "internal" variant.  This
is needed to share some functionality across modules inside the
`muste` package.  These modules are not exposed and so therefore
cannot be accessed from `muste-ajax`.  The module `Muste` re exports
all non internal modules and defines a pair of symbols itself.  To see
what is exposed from the package `muste` please refer to the Haddock
documentation (see `README.md`).

I plan to maintain a change log and a back log.  The back log contains
currently contains various ideas I got while cleaning things up.

The internal dependency graph of `muste` looks like this
(transitive dependencies are not shown as direct edges):

[muste.pdf](https://github.com/MUSTE-Project/MULLE/files/2197290/muste.pdf)

Addressing specific questions/remarks
---

> 5 actually (at least): TTree, LTree, MetaTTree, CostTree, GFAbsTree

The only representations being exported now are `TTree` and
`CostTree`.

> GFAbsTree is used by PGF, but should only be used in an PGF
> interface module (e.g., Tree/Internal.hs). Currently it's also used
> in Main.hs

Currently the `muste-ajax` component continues to use `PGF` directly
for reading `.pgf`-files.

> There is also LinToken and Linearization (in Muste.hs) - should be merged into one.

These have not been merged, but only `LinToken` is exposed.

> > Which functions fit best where?
> 
> linearization functions can be moved to a new module (e.g. Muste.Linearize): LinToken, Linearization, linearizeTree (are defined in Muste.hs)

I have created a new module `Muste.Linearization` with the following
members: `LinToken`, `Context`, `buildContext` and `linearizeTree`.

> the Ajax backend should only refer to functions exported by Muste.hs: now Ajax.hs imports Muste.Tree, Main-CGI.hs imports Muste.Grammar, and Database.hs and Protocol.hs imports both Muste.Tree, Muste.Grammar and PGF
> is there a need for Muste.Grammar.Internal? and Muste.Tree.Internal?

`muste-ajax` can only access exposed members. Some functionality
*inside* `muste` needs access to members defined elsewhere (at least
ATM). By having an "internal" variant of a module we can avoid
exposing these members.

> perhaps Database.hs should be moved to the muste-backend??

I'd prefer wrapping the functionality of parsing a `.pgf`-files.

> Also, Database.hs contains explicit example trees

These trees are needed for initializing the database. I have moved all
"dumb data" into a separate module.

> only Muste.Grammar (or Muste.Grammar.Internal) should use the PGF module -- everything that the other modules need should be exported by Muste.Grammar

Currently both the modules `Muste` and `Muste.Linearization.Internal`
need access to `ttreeToGFAbsTree`.  I think this is strictly needed to
break a dependency cycle between `Linearization` and `Grammar`.
Furthermore the test-suite accesses `Muste.Grammar.Internal`.  I
haven't looked at this in depth.

> Muste.Feat currently defines the types Grammar, Rule, FunType --
> this should be done by Muste.Grammar (and the functions getAllRules,
> getRuleType)

Done

> Muste.Feat currently defines TTree -- this should be in Muste.Tree
> instead

Done

> 
> functions in Muste.Prune that should be in other files:
> 
> getFunctions, sameRoot, getMetas, treeDiff, treeIsRecursive, ... (move to Muste.Tree)

Moving `getFunctions` and `getMetas` will require me to move the
definition of `Rule` to `Tree` as well.  Is this desired?  I want to
point out that `Muste.Prune` can be made an internal module (i.e. it
will not be exposed to other packages).

> editDistance, isSubList, areDisjoint, noDuplicates, ... (move to Muste.Common)

Done
